projects/interfaces/main.go

```diff
@@ -0,0 +1,128 @@
+package main
+
+import (
+	"io"
+	"unicode"
+)
+
+// ---------- Implementing our own bytes.Buffer
+
+// the original Buffer type definition:
+//
+//	type Buffer struct {
+//		buf      []byte // contents are the bytes buf[off : len(buf)]
+//		off      int    // read at &buf[off], write at &buf[len(buf)]
+//		lastRead readOp // last read operation, so that Unread* can work correctly.
+//	}
+type OurByteBuffer struct {
+	buf []byte
+}
+
+// the original Buffer Bytes() method signature:
+// func (b *Buffer) Bytes() []byte
+func (b *OurByteBuffer) Bytes() []byte {
+	// simply return the buffer
+	return b.buf
+}
+
+// the original Buffer Write() method signature:
+// func (b *Buffer) Write(p []byte) (n int, err error)
+func (b *OurByteBuffer) Write(p []byte) (int, error) {
+	// Q: Why is no error returned from append(?)
+	// A: The append function in Go doesn't generally return an error because it performs a safe operation that cannot fail in most cases.
+	// The reason for this is related to how Go handles memory allocation and slice operations.
```

Comment:
When we say Go treats allocation as a "safe operation", we means the operation can't fail (or, in this case, the language pretends it cannot fail).

In general, if something can fail, it should return an error so that the caller can handle the failure.

If your system is out of memory, it's likely there's nothing you can actually do about this (and it's even more likely that _the specific code trying to append to a slice_ can't do anything useful about it), so rather than _every_ operation which may allocate memory returning an error, Go pretends that it can't fail (and if it can, your whole program is going to fail too), so panics.